### PR TITLE
Improve BCP type checks

### PR DESCRIPTION
**Pull request by webmaster128**
Required to get https://github.com/iov-one/ponferrada/issues/393 done

I decided to fix this here in IOV-Core in order to make the use of this library more convenient for everybody.

### CODE REVIEW

File: packages/iov-bcp/src/transactions.spec.ts [hunk 1]
```diff
@@ -1,6 +1,152 @@
...
+  isUint8Array,
+  PubkeyBundle,
+  PubkeyBytes,
+} from "./transactions";
+
+const { fromHex } = Encoding;
 
 describe("transactions", () => {
+  describe("isNonNullObject", () => {
+    it("returns true for objects", () => {
```
Comment (willclarktech) [inline packages/iov-bcp/src/transactions.spec.ts]: "including arrays" to help with clarity here?

File: packages/iov-bcp/src/transactions.spec.ts [hunk 2]
```diff
@@ -1,6 +1,152 @@
...
+    });
+
+    it("returns false for a bunch of other stuff", () => {
+      expect(isPubkeyBundle("abc")).toEqual(false);
+      expect(isPubkeyBundle({})).toEqual(false);
+      expect(isPubkeyBundle(null)).toEqual(false);
+      expect(isPubkeyBundle(undefined)).toEqual(false);
+      expect(isPubkeyBundle(fromHex("aabb"))).toEqual(false);
+
+      const missingAlgo: Omit<PubkeyBundle, "algo"> = {
```
Comment (willclarktech) [inline packages/iov-bcp/src/transactions.spec.ts]: 👌

File: packages/iov-bcp/src/transactions.spec.ts [hunk 3]
```diff
@@ -1,6 +1,152 @@
...
+        data: fromHex("aabbccdd") as PubkeyBytes,
+      };
+      expect(isPubkeyBundle(missingAlgo)).toEqual(false);
+
+      const missingData: Omit<PubkeyBundle, "data"> = {
+        algo: Algorithm.Ed25519,
+      };
+      expect(isPubkeyBundle(missingData)).toEqual(false);
+    });
```
Comment (willclarktech) [inline packages/iov-bcp/src/transactions.spec.ts]: Expected behaviour with additional properties?

File: packages/iov-bcp/src/transactions.spec.ts [hunk 4]
```diff
@@ -1,6 +1,152 @@
...
+          data: fromHex("aabbccdd") as PubkeyBytes,
+        },
+      };
+      expect(isIdentity(missingChainId)).toEqual(false);
+
+      const missingPubkey: Omit<Identity, "pubkey"> = {
+        chainId: "foobar" as ChainId,
+      };
+      expect(isIdentity(missingPubkey)).toEqual(false);
+    });
```
Comment (willclarktech) [inline packages/iov-bcp/src/transactions.spec.ts]: As above

File: packages/iov-bcp/src/transactions.ts [hunk 5]
```diff
@@ -3,6 +3,33 @@ import { As } from "type-tagger";
 
 import { Hash, Preimage } from "./atomicswaptypes";
 
+/**
+ * Checks is data is a non-null object (i.e. matches the TypeScript object type)
```
Comment (willclarktech) [inline packages/iov-bcp/src/transactions.ts]: "Checks if"

File: packages/iov-bcp/src/transactions.ts [hunk 6]
```diff
@@ -3,6 +3,33 @@ import { As } from "type-tagger";
...
+ *
+ * Only used package-internal.
```
Comment (willclarktech) [inline packages/iov-bcp/src/transactions.ts]: "internally to this package"

File: packages/iov-bcp/src/transactions.ts [hunk 7]
```diff
@@ -3,6 +3,33 @@ import { As } from "type-tagger";
...
+ */
+export function isNonNullObject(data: unknown): data is object {
+  return typeof data === "object" && data !== null;
+}
+
+/** Checks if data is an Uint8Array. Note: Buffer is treatet as not a Uint8Array */
```
Comment (willclarktech) [inline packages/iov-bcp/src/transactions.ts]: "treated"

File: packages/iov-bcp/src/transactions.ts [hunk 8]
```diff
@@ -15,13 +42,12 @@ export interface PubkeyBundle {
   readonly data: PubkeyBytes;
 }
 
-export function isPubkeyBundle(data: any): data is PubkeyBundle {
+export function isPubkeyBundle(data: unknown): data is PubkeyBundle {
```
Comment (willclarktech) [inline packages/iov-bcp/src/transactions.ts]: 👌

File: packages/iov-bcp/src/transactions.spec.ts (hunk 1 above)
Comment (webmaster128) [reply packages/iov-bcp/src/transactions.spec.ts]: Hah! I bend the tests to the behavior of JavaScript. My original expectation was `expect(isNonNullObject([])).toEqual(false);`.

Looking at the definition of the `object` type, an Array is an object:

> object is a type that represents the non-primitive type, i.e. anything that is not number, string, boolean, symbol, null, or undefined.

https://www.typescriptlang.org/docs/handbook/basic-types.html#object

So 👍 for making that explicit

**Comment by webmaster128** [comment]
All comments implemented

**Pull request merged**